exp/controllers/ocimachinepool_controller.go

```diff
@@ -0,0 +1,384 @@
+		machinePoolScope.SetFailureReason(capierrors.CreateMachineError)
+		machinePoolScope.SetFailureMessage(errors.Errorf("Instance Pool status %q is unexpected", instancePool.LifecycleState))
+		r.Recorder.Eventf(machinePoolScope.OCIMachinePool, corev1.EventTypeWarning, "ReconcileError",
+			"Instance pool has invalid lifecycle state %s", instancePool.LifecycleState)
+		return reconcile.Result{}, errors.New(fmt.Sprintf("instance pool has invalid lifecycle state %s", instancePool.LifecycleState))
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *OCIMachinePoolReconciler) reconcileDelete(ctx context.Context, machinePoolScope *scope.MachinePoolScope) (_ ctrl.Result, reterr error) {
+	machinePoolScope.Info("Handling deleted OCIMachinePool")
+
+	// Find existing Instance Pool
+	instancePool, err := machinePoolScope.FindInstancePool(ctx)
+	if err != nil {
+		conditions.MarkUnknown(machinePoolScope.OCIMachinePool, infrav1exp.InstancePoolReadyCondition, infrav1exp.InstancePoolNotFoundReason, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	if instancePool == nil {
+		machinePoolScope.V(2).Info("Unable to locate instance pool", "id", machinePoolScope.OCIMachinePool.Spec.ProviderID)
+		r.Recorder.Eventf(machinePoolScope.OCIMachinePool, corev1.EventTypeNormal, infrav1exp.InstancePoolNotFoundReason, "Unable to find matching instance pool")
+	} else {
+		switch instancePool.LifecycleState {
+		case core.InstancePoolLifecycleStateTerminating:
+			// Instance Pool is already deleting
+			machinePoolScope.OCIMachinePool.Status.Ready = false
+			conditions.MarkFalse(machinePoolScope.OCIMachinePool, infrav1exp.InstancePoolReadyCondition, infrav1exp.InstancePoolDeletionInProgress, clusterv1.ConditionSeverityWarning, "")
+			r.Recorder.Eventf(machinePoolScope.OCIMachinePool, corev1.EventTypeWarning, "DeletionInProgress", "Instance Pool deletion in progress: %s - %s", instancePool.DisplayName, instancePool.Id)
+			machinePoolScope.Info("Instance Pool is already deleting", "displayName", instancePool.DisplayName, "id", instancePool.Id)
+		case core.InstancePoolLifecycleStateTerminated:
+			machinePoolScope.Info("Instance Pool is already deleted", "displayName", instancePool.DisplayName, "id", instancePool.Id)
+		default:
+			machinePoolScope.Info("Deleting instance pool", "id", instancePool.Id, "lifecycleState", instancePool.LifecycleState)
+			req := core.TerminateInstancePoolRequest{InstancePoolId: instancePool.Id}
+			if _, err := machinePoolScope.ComputeManagementClient.TerminateInstancePool(ctx, req); err != nil {
+				r.Recorder.Eventf(machinePoolScope.OCIMachinePool, corev1.EventTypeWarning, "FailedDelete", "Failed to delete instance pool %q: %v", instancePool.Id, err)
+				return ctrl.Result{}, errors.Wrap(err, "failed to delete instance pool")
+			}
+		}
+	}
+
+	instanceConfigurationId := machinePoolScope.OCIMachinePool.Status.InstanceConfigurationId
```

Comment:
we should save the id in the spec, not the status. status will be deleted during move which will make us fail reconciliation

Reply:
please do for all ids